### PR TITLE
Release pinned buffers correctly for WinHttpHandler, WinHttpWebSocket

### LINKED ISSUE #2508: Release pinned buffers at correct time in WinHttpResponseStream, WinHttpWebSocket
This is a potential bug: pinned buffers must not be released until it is guaranteed that no further operations will be made that use those buffers.

### PR DESCRIPTION
**Pull request by davidsh**
Both WinHttpHandler and WinHttpWebSocket use pinned buffers for receiving and sending data. The pinned GCHandle was being freed during the dispose of the WinHttpHandler or WinHttpWebSocket objects. The problem is that the buffer should not be unpinned until the async operation is done. It's possible for any async operation to be in flight (and where native code uses the buffer) when the object is explicitly or implicitly disposed.

This fix moves the ownership of the pinned GCHandle to their respective state objects (WinHttRequestState, WinHttpWebSocketState). These objects are strongly rooted and are not released until all async operations are completed and the WinHTTP handles are fully closed. That makes these state objects a safe place to own and free the pinned buffers.

Fixes #2508 and #6142

### CODE REVIEW

**Comment by davidsh** [comment]
@stephentoub @CIPop PTAL

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs [hunk 1]
```diff
@@ -147,7 +163,17 @@ private void Dispose(bool disposing)
 
             if (_operationHandle.IsAllocated)
             {
+                // This method only gets called when the WinHTTP request handle is fully closed and thus all
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs]: Is there anything we can assert to ensure that?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs [hunk 2]
```diff
@@ -123,6 +123,22 @@ public void ClearSendRequestState()
         public long? ExpectedBytesToRead { get; set; }
         public long CurrentBytesRead { get; set; }
 
+        // TODO (Issue 2505): temporary pinned buffer caches of 1 item. Will be replaced by PinnableBufferCache.
+        private GCHandle _cachedReceivePinnedBuffer = default(GCHandle);
+
+        public void PinReceiveBuffer(byte[] buffer)
+        {
+            if (!_cachedReceivePinnedBuffer.IsAllocated || _cachedReceivePinnedBuffer.Target != buffer)
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs]: What guarantees do we have that there isn't still an outstanding operation using the old target?  Maybe there's something we can assert, about there not being any outstanding operations or something?

**Comment by stephentoub** [comment]
LGTM

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs (hunk 2 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs]: The WinHttpRequestState object relies on the caller (and there is only one caller) to assert that.  In particular, the only async operation possible on this pinned receive buffer is from WinHttpResponseStream.ReadAsync().  There is already a check in that public API surface to reject calls to it if the previous ReadAsync is in progress.

Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs]: Ok.

Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs]: FWIW #2501 will ensure more consistency checks as well.

**Pull request merged**